PR: Use QRawFont and draw PainterPath of Glyph for animation to prevent tremulous spinning icons

kumattau opened the pull request:
Maybe #39 can be fixed using QRawFont and draw PainterPath of Glyph.

Compared to QFont, QRawFont is available without registering to font database,
So this approach has

Pros:
* no tremulous spinning
* no family name collision (font patch is not required)

Cons:
* cannot support color-font (PainterPath of Glyph does not have color)
* cannot support font() API (this requires QFont registered to font database)
* I tests it on ubuntu only. I cannot check this approach works on other platform or not

If qtawesome switches QFont to QRawFont, font patch in the update_font script ( #194 ) may not be needed.

How do you think, @dalthviz  ?

Note: current code is example, so it needs refactoring, optimization, etc.

Edit:

Fixes #39

dalthviz commented:
Hi @kumattau thanks for checking into this! Testing this locally seems like this also works for Windows :+1:

![example](https://user-images.githubusercontent.com/16781833/151224184-e24e1900-2e9e-4f04-bc9c-87a4804cbfd2.gif)

Regarding the pros and cons maybe we could use `QRawFont` only when needed as you are doing here but still registering the font to still have access to the font API if needed?

Just in case what do you think @ccordoba12 ?

kumattau commented:
Thank you for testing @dalthviz

On windows, "Stack, offset" bled quite a bit. this is because QPainterPath does not have hinting.
So, I changed it to use QPainterPath only for animations to keep rendering quality.

On my ubuntu, example.py is here.
Could you compare "Stack, offset" and "Stack, offset, draw path" on Windows ?

![Screenshot from 2022-01-29 00-45-56](https://user-images.githubusercontent.com/171798/151577639-b84705a2-ffc9-4b16-876c-16e258336ec0.png)

The drawing method can be selected by "draw" option.

* path: do painter.drawPath()
* text: do painter.drawText()
* auto: do painter.drawPath when animation, otherwise do painter.drawText() (default)

kumattau commented:
@dalthviz, I finished the update. This PR is ready for review.

dalthviz commented:
Checking the `example.py` on Windows this is what I'm getting:

![imagen](https://user-images.githubusercontent.com/16781833/151604441-ba633409-5047-4143-9476-a095458ed7f8.png)

dalthviz commented on example.py:
```diff
@@ -114,6 +114,14 @@ def __init__(self):
                                     {'scale_factor': 0.8}])
         saveall_button = QtWidgets.QPushButton(saveall, 'Stack, offset')
 
+        # Stack and offset icons, draw path
+        saveall_path = qta.icon('fa5.save', 'fa5.save',
+                           options=[{'scale_factor': 0.8,
+                                     'offset': (0.2, 0.2),
+                                     'color': 'gray', 'draw': 'path'},
+                                    {'scale_factor': 0.8, 'draw': 'path'}])
+        saveall_button_path = QtWidgets.QPushButton(saveall_path, 'Stack, offset, draw path')
+
```
I think we should add this to the docs and probably update the GIF for `example.py`. Maybe adding to the docs some info regarding the new `draw` option

dalthviz commented on qtawesome/iconic_font.py:
```diff
@@ -173,7 +173,47 @@ def _paint_icon(self, iconic, painter, rect, mode, state, options):
+        elif draw == 'path':
+            draw_path = True
+        else:
+            # Use QPainterPath when animation
+            # to fix tremulous spinning icons #39
```
Improve a little bit the comment here
```suggestion
            # Use QPainterPath when setting an animation
            # to fix tremulous spinning icons.
            # See #39
```

dalthviz commented on qtawesome/iconic_font.py:
```diff
@@ -270,14 +312,23 @@ def hook(obj):
-                id_ = QFontDatabase.addApplicationFontFromData(QByteArray(font_data.read()))
+                data = font_data.read()
+                id_ = QFontDatabase.addApplicationFontFromData(data)
+                # QRawFont() requires pixelSize, but it is unknown at this time.
+                # So, first create QRowfont with 2048, which is the size of
+                # almost TrueType fonts and reset actual size after that.
+                # Note: When TrueType fonts, pixelSize can be specified up to 16384.
```
A little improvement to the comment here:
```suggestion
                # `QRawFont()` requires a `pixelSize` value, but it is unknown at this point.
                # To workaround that, first we create a `QRawFont` with 2048 as `pixelSize` 
                # (usual size for TrueType fonts) and reset to the actual size after that.
                # Note: For TrueType fonts, `pixelSize` can be specified up to 16384.
```

dalthviz reviewed (commented):
Thanks @kumattau for working on this! I left some suggestions regarding the comments in the code and also maybe we need to also update the docs (at least the example code in this PR) but otherwise this LGTM :+1:

kumattau commented:
Thank you for review @dalthviz, I updated comment.
I've come up with a better way to do it and will experiment with it. Please wait a little longer.

kumattau commented:
Hi, @dalthviz 

> I've come up with a better way to do it and will experiment with it.

I try to drawImage a glyph of QRawfont.
On my ubuntu, the drawing quality is same as original drawText with QFont.
Could you try example.py on Windows and feedback me ?

If the drawImage is okey, qtawesome does not require for QFont except for font() API.
I think I want not to patch fonts if possible, so, it is the first step of the use of patch-free font.

dalthviz commented:
Hi again @kumattau, this is what I'm seeing on Windows with the latest changes:

![imagen](https://user-images.githubusercontent.com/16781833/152024403-875efad8-cf16-4dc4-9247-d6d5d3176e9e.png)

So, indeed the icon drawing locks cleaner :+1: 

However seems like the [new code is not compatible with Qt <5.15](https://forum.qt.io/topic/121460/qimage-setalphachannel-obsolete-or-not/3). The tests are failling here since it our CI uses Qt 5.9.7 and locally I tried first testing with an env that had Qt 5.12.2 and the same error appeared. Could be possible to have a fallback in case `QImage.setAlphaChannel` doesn't exists in the Qt/PyQt/PySide version installed?

kumattau commented:
@dalthviz 

> Could be possible to have a fallback in case QImage.setAlphaChannel doesn't exists in the Qt/PyQt/PySide version installed?

Yes. I added fallback code and pushed it.
I will try to refactor the code and add three draw options (path, image, text).

kumattau commented:
Hi, @dalthviz 

> cannot support color-font (PainterPath of Glyph does not have color)

I found another issue.
QRawFont requires that the font has glyph.
So, for example, NotoColorEmoji.ttf, which does not have glyphs, cannot be rendered using QRawFont.

https://github.com/fonttools/fonttools/issues/949#issuecomment-299148957

qtawesome cannot switch the drawing method from `drawPath()` to the original `drawText()` if the font does not have glyphs.
Because QRawFont have `bool supportsCharacter(char)` method to check if the font has a glyph that corresponds to the given character, but this always returns True on ubuntu.

And the following error message output to the console if a glyph is missing, but this message cannot stopped.

> load glyph failed err=24

Should qtawesome support a font which does not have glyphs ?

dalthviz commented:
Hi again @kumattau !

> Should qtawesome support a font which does not have glyphs ?

I would say yes. Probably we need to find a reliable way to detect if the font does not have the needed glyphs, show some sort of warning letting users know that we were unable to use the specified config to draw the icon and then fallback to use the `drawText`/``text` option.

Maybe there are other ways to check for the character support more reliable than `QRawFont.supportsCharacter` ?

kumattau commented:
Hi @dalthviz 

> I would say yes.

I got it.

> I would say yes. Probably we need to find a reliable way to detect if the font does not have the needed glyphs, show some sort of warning letting users know that we were unable to use the specified config to draw the icon and then fallback to use the drawText/``text` option.

I think fallback is possible.
If glyph is missing, the PainterPath is empty, so I can implement the fallback to drawText when PainterPath is emtpy.
The biggest issue at the moment is that I can't stop the following error messages.

> load glyph failed err=24

It is quite annoying because it is always output when glyph is missing.

I will do some more research on this matter.

dalthviz commented:
Hi again @kumattau , hope everything is going well! Since we will try to release v1.2.0 soon, just to know if we can make this part of the release a couple of questions: Did you find any other info that could help with the missing elements here? and also, would you like to finish this or is okay for me or someone else to try finish it? Let us know!

kumattau commented:
@dalthviz Thank you for follow-up and I'm very sorry not to response.

> I will do some more research on this matter.

The investigation has not progressed but I will reboot this PR soon (in this week or next week).
If my schedule is not suite the release, I'm no problem for you to try finish it at all.

dalthviz commented:
Thanks for the prompt response @kumattau ! If you still want to finish this at some point then no problem, happy to hear that :)

It would be nice to merge this for the next release v1.2.0 (which will be done probably mid-next week) but do not worry about the time, just knowing that this could be finished in the future and you are willing to keep checking on it are really good news :+1:

kumattau commented:
@dalthviz 

I rebased master and fixed "load glyph failed err=24" problem by fall backing draw=text
The draw option accepts text/path/glyphrun/image. If not specified, draw=path is used for non-animation and draw=text is used for animation.
example.py accepts global_default options (key=val format) for test.

In my linux, it is better that draw=image is used for non-animation and draw=text is used for animation.
Could you please try it in windows ?

kumattau commented:
Looking to spin, I still think draw=path would be better when animation.

dalthviz commented:
Thanks for the quick update @kumattau ! Doing some testing on Windows:

### Default options (no args) - `python example.py`

![default](https://user-images.githubusercontent.com/16781833/195893631-791b062f-df2d-4861-a136-2e21333cdad0.gif)

### `draw=image` option - `python example.py draw=image`

![draw_image](https://user-images.githubusercontent.com/16781833/195893923-4547ad0a-3ad7-45c8-9fb2-c361bdf90444.gif)

### `draw=path` option - `python example.py draw=path`

![draw_path](https://user-images.githubusercontent.com/16781833/195894410-b04889a2-bee8-4411-b922-be9d8270d03c.gif)

### `draw=text` option - `python example.py draw=image`

![draw_text](https://user-images.githubusercontent.com/16781833/195894752-88a4c920-1464-410c-a255-b25384af1f14.gif)

Checking then seems like the `draw=path` should be the default when animation is requested and `draw=image` when not

dalthviz commented on qtawesome/iconic_font.py:
```diff
@@ -178,45 +179,94 @@ def _paint_icon(self, iconic, painter, rect, mode, state, options):
+            # Use QPainterPath when setting an animation
+            # to fix tremulous spinning icons.
+            # See #39
+            draw = 'path' if animation is not None else 'text'
```
Following our findings I think here we should change `text` for `image`:

```suggestion
            draw = 'path' if animation is not None else 'image'
```

kumattau replied:
Looking at Material Design and Rotate, image seems not to be good.

kumattau replied:
As https://github.com/spyder-ide/qtawesome/pull/202#issuecomment-1279741225, I think `text` or `glyphrun` may be  better for non-animation on linux and windows.

dalthviz replied:
You are right! I missed checking that and just focused on the remix icon that looked sharper with `draw=image` 😅

dalthviz commented:
Besides the suggestion above I think the only missing element is updating the `example.py` gif and documentation (adding the new draw option and also the possiblity to pass options to the `example.py` script) but I can take care of that in a follow-up PR before doing the release :+1:

dalthviz commented on qtawesome/iconic_font.py:
```diff
@@ -178,45 +179,94 @@ def _paint_icon(self, iconic, painter, rect, mode, state, options):
+        if draw not in ('text', 'path', 'glyphrun', 'image'):
+            # Use QPainterPath when setting an animation
+            # to fix tremulous spinning icons.
+            # See #39
```
Let change the comment to follow other comments pointing to issues:
```suggestion
            # See spyder-ide/qtawesome#39
```

kumattau commented:
@dalthviz 

How about draw=glyphrun on windows ?

On linux, The quality of draw=glyphrun seems to be same as draw=text.

`glyphrun` can use rawfont (in other words, does not requires the installation of font) but cannot render emoji (without glyph).

dalthviz commented:
Checking `draw=glyphrun` on Windows I got this:

![image](https://user-images.githubusercontent.com/16781833/196008299-6e93e27b-0fec-40dc-873e-a1d955f1e3d1.png)

Looking more in detail the different draw options I would say that `glyphrun` looks quite similar to `text` and also as you said for some fonts using `image` doesn't look quite good, so lets leave the default draw option to be `text` for non-animated icons :+1:

Thanks again @kumattau for the work here and checking all the draw options! I think this is ready to be merged then, right? Or maybe do you want to check something else on Windows? If is ready to be merged just mark the PR as `Ready for review` and I will merge it :+1:

kumattau commented:
@dalthviz 

My work has been finished.
The commits keep not to be squashed.
If you want to simplify the PR, please squash and merge it.

dalthviz reviewed (approved):
Thanks for all the work done here @kumattau !
